agent/s3.go

```diff
@@ -107,35 +84,51 @@ func webIdentityRoleProvider(sess *session.Session) *stscreds.WebIdentityRolePro
 }
 
 func newS3Client(l logger.Logger, bucket string) (*s3.S3, error) {
-	region, err := awsS3RegionFromEnv()
-	if err != nil {
-		// Fallback region guess
-		region = "us-east-1"
-	}
+	var sess *session.Session
 
-	// Using the guess region, construct a session and ask that region where the
-	// bucket lives
-	sess, err := awsS3Session(region)
-	if err != nil {
-		return nil, err
-	}
-	bucketRegion, err := s3manager.GetBucketRegion(aws.BackgroundContext(), sess, bucket, region)
-	if err != nil {
-		return nil, err
-	}
+	regionHint := os.Getenv(regionHintEnvVar)
+	if regionHint != "" {
+		// If there is a region hint provided, we use it unconditionally
+		session, err := awsS3Session(regionHint)
+		if err != nil {
+			return nil, fmt.Errorf("Could not load the AWS SDK config (%v)", err)
+		}
 
-	// Construct the final session for the bucket region
-	sess, err = awsS3Session(bucketRegion)
-	if err != nil {
-		return nil, err
+		sess = session
+	} else {
+		// Otherwise, use the current region (or a guess) to dynamically find
+		// where the bucket lives.
+		region, err := awsRegion()
+		if err != nil {
+			region = "us-east-1"
+		}
+
+		l.Debug("Discovered current region as %q\n", region)
+
+		// Using the guess region, construct a session and ask that region where the
+		// bucket lives
+		session, err := awsS3Session(region)
+		if err != nil {
+			return nil, fmt.Errorf("Could not load the AWS SDK config (%v)", err)
+		}
+
+		bucketRegion, bucketRegionErr := s3manager.GetBucketRegion(aws.BackgroundContext(), sess, bucket, region)
+		if bucketRegionErr == nil && bucketRegion != "" {
+			l.Debug("Discovered %q bucket region as %q\n", bucket, bucketRegion)
+			session.Config.Region = &bucketRegion
```

Comment:
Needs testing to verify that this can be set after the session is constructed

Reply:
Verified using a local agent that this does correctly re-configure the session and subsequent client to use the discovered region.